### PR TITLE
Allow customising the default ordering of Snippets IndexView

### PR DESCRIPTION
**Pull request by laymonage**
Part of wagtail/rfcs#85. Incorporates #10275.

This PR adds an `ordering` attribute and `get_ordering(request)` method to the `SnippetViewSet` to mirror `ModelAdmin.get_ordering(request)`. On the Snippets `IndexView.get_ordering()`, we first try to call `SnippetViewSet.get_ordering(self.request)`. If the result is `None` (the default), the original implementation in the generic `IndexView` is used, where it checks the request's `GET` parameters before resorting to the default ordering.

In addition, a private `IndexView._get_ordering()` method is extracted (and this is the one that's actually used before calling `QuerySet.order_by()`). This method only acts to normalise the result of `get_ordering()` to ensure that it results in a tuple (so we can do `order_by(*ordering)`) and to ensure that we handle the nulls first/last issue with the `_updated_at` annotated column.

**Note:** I don't know why we have a `default_ordering` attribute on the generic `IndexView`. It extends Django's `BaseListView`, so it has the `ordering` attribute. Why not use it? 🤔

Anyway, I also made it so that the code looks for `self.ordering` before falling back to the existing `self.default_ordering`.

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [x] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

Continuing the test example in #10275:

```py
from wagtail.snippets.models import register_snippet
from wagtail.snippets.views.snippets import SnippetViewSet


class PersonViewSet(SnippetViewSet):
    icon = "user"
    list_per_page = 2
    chooser_per_page = 1
    admin_url_namespace = "some_namespace"
    base_url_path = "deep/within/the/admin"
    chooser_admin_url_namespace = "my_chooser_namespace"
    chooser_base_url_path = "choose/wisely"
    # this will generate exact filters
    list_filter = ["job_title", "first_name"]
    # alternatively, this will generate filters with lookups
    list_filter = {
        "job_title": ["contains"],
        "first_name": ["contains"],
        "first_published_at": ["date"],
    }
    ordering = "job_title"
    # alternatively
    ordering = ["job_title", "last_name", "first_name"]
    # alternatively
    def get_ordering(self, request):
        return ["job_title", "last_name", "first_name"]  # or some other custom logic

    def get_queryset(self, request):
        return self.model._default_manager.all().exclude(job_title__contains="[HIDDEN]")


register_snippet(Person, viewset=PersonViewSet)
```

The listing view will sort by the specified ordering initially. If any of the columns are sortable, they'll continue to work normally.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

File: wagtail/admin/views/generic/models.py [hunk 1]
```diff
@@ -327,17 +319,48 @@ def get_add_url(self):
...
+        default_ordering = self.ordering or self.default_ordering
+        if default_ordering:
+            if isinstance(default_ordering, str):
+                default_ordering = (default_ordering,)
+            orderings |= set(default_ordering)
         return orderings
 
     def get_ordering(self):
-        ordering = self.request.GET.get("ordering", self.default_ordering)
+        default_ordering = self.ordering or self.default_ordering
```
Comment (laymonage) [inline wagtail/admin/views/generic/models.py:337]: I don't know why we have a default_ordering attribute on the generic IndexView. It extends Django's BaseListView, so it has the ordering attribute. Why not use it? 🤔

File: wagtail/admin/views/generic/models.py [hunk 2]
```diff
@@ -327,17 +319,48 @@ def get_add_url(self):
...
+        updated_at_asc = models.F("_updated_at").asc(nulls_first=True)
+        updated_at_desc = models.F("_updated_at").desc(nulls_last=True)
+        ordering = (
+            updated_at_asc
+            if order == "_updated_at"
+            else updated_at_desc
+            if order == "-_updated_at"
+            else order
+            for order in ordering
+        )
```
Comment (laymonage) [inline wagtail/admin/views/generic/models.py:363]: The logic has been updated to handle the ordering as a list/tuple to make sure we handle it if multiple ordering fields have been specified (either by the default ordering, or multiple `GET` parameters, even though the latter isn't officially supported via the UI). The previous logic (deleted in `get_queryset()` above) only handles the case where only a single ordering field is specified.

**Review by gasman — CHANGES_REQUESTED** [review]
I have to say, I'm finding the "`get_ordering` as a callable" mechanism (and thus the need for a separate `_get_ordering` method) pretty confusing. If I see a `get_ordering` method with a request parameter, that would suggest to me that it's responsible for looking at the `ordering` URL parameter and acting on that, but apparently here it's only setting the _default_ ordering and the URL parameter still takes precedence over that.

Can we simplify this PR to just support the `ordering` attribute on the viewset, rather than `get_ordering` too? If people have really esoteric requirements for varying the default order per request, I think it's fair game to require them to override the relevant views and pass those to the viewset class.

**Comment by laymonage** [comment]
Thanks @gasman! I'm glad that we're not aiming for 1:1 parity with ModelAdmin here, as I was hesitant to pass all those callables. I've simplified the PR to only pass the `ordering` attribute as the `default_ordering` on the view.

Not sure why the readthedocs build failed, but it fails due to some git issue and not because of the content. It builds fine on my machine.

File: docs/topics/snippets.md [hunk 3]
```diff
@@ -575,7 +575,7 @@ Similar URL customisations are also possible for the snippet chooser views throu
 
 The {attr}`~wagtail.snippets.views.snippets.SnippetViewSet.list_display` attribute can be set to specify the columns shown on the listing view. To customise the number of items to be displayed per page, you can set the {attr}`~wagtail.snippets.views.snippets.SnippetViewSet.list_per_page` attribute (or {attr}`~wagtail.snippets.views.snippets.SnippetViewSet.chooser_per_page` for the chooser listing).
 
-To customise the base queryset for the listing view, you could override the {meth}`~wagtail.snippets.views.snippets.SnippetViewSet.get_queryset` method.
+To customise the base queryset for the listing view, you could override the {meth}`~wagtail.snippets.views.snippets.SnippetViewSet.get_queryset` method. Additionally, the {attr}`~wagtail.snippets.views.snippets.SnippetViewSet.ordering` attribute or the {meth}`~wagtail.snippets.views.snippets.SnippetViewSet.get_ordering` method can be used to specify the default ordering of the listing view.
```
Comment (gasman) [inline docs/topics/snippets.md:578]: The mention of `get_ordering` should be removed here - will do that when I merge.

**Review by gasman — APPROVED** [review]

**Comment by gasman** [comment]
Merged in 3965ceedfa74647e4083e21b87a339704239b75c

**Pull request closed without merging**